### PR TITLE
Simplify data verification with HashReader.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Verify() was being called by caller after the data
has been successfully read after io.EOF. This disconnection
opens a race under concurrent access to such an object.
Verification is not necessary outside of Read() call,
we can simply just do checksum verification right inside
Read() call at io.EOF.

## Motivation and Context
Simplification of usage

## How Has This Been Tested?
Manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/xl-v1-object.go [hunk 1]
```diff
@@ -531,8 +530,8 @@ func (xl xlObjects) PutObject(bucket string, object string, data *HashReader, me
 		// Hint the filesystem to pre-allocate one continuous large block.
 		// This is only an optimization.
 		if curPartSize > 0 {
-			pErr := xl.prepareFile(minioMetaTmpBucket, tempErasureObj, curPartSize, storage.disks, xlMeta.Erasure.BlockSize, xlMeta.Erasure.DataBlocks)
-			if pErr != nil {
+			if pErr := xl.prepareFile(minioMetaTmpBucket, tempErasureObj, curPartSize, storage.disks,
```
Comment (balamurugana) [inline cmd/xl-v1-object.go]: why do need to change this line?

File: cmd/xl-v1-object.go [hunk 2]
```diff
@@ -113,7 +112,7 @@ func (xl xlObjects) CopyObject(srcBucket, srcObject, dstBucket, dstObject string
 		pipeWriter.Close() // Close writer explicitly signalling we wrote all data.
 	}()
 
-	objInfo, err := xl.PutObject(dstBucket, dstObject, NewHashReader(pipeReader, length, metadata["etag"], ""), metadata)
+	objInfo, err := xl.PutObject(dstBucket, dstObject, NewHashReader(pipeReader, length, "", ""), metadata)
```
Comment (balamurugana) [inline cmd/xl-v1-object.go]: why etag is not passed?

File: cmd/test-utils_test.go [hunk 3]
```diff
@@ -1975,6 +1975,7 @@ func ExecObjectLayerTest(t TestErrHandler, objTest objTestType) {
 	if err != nil {
 		t.Fatalf("Initialization of object layer failed for single node setup: %s", err)
 	}
+
```
Comment (balamurugana) [inline cmd/test-utils_test.go:1987]: My recommendation is to avoid unrelated diffs

File: cmd/object-handlers.go [hunk 4]
```diff
@@ -287,6 +287,9 @@ func (api objectAPIHandlers) HeadObjectHandler(w http.ResponseWriter, r *http.Re
 // Extract metadata relevant for an CopyObject operation based on conditional
 // header values specified in X-Amz-Metadata-Directive.
 func getCpObjMetadataFromHeader(header http.Header, defaultMeta map[string]string) (map[string]string, error) {
+	// Make sure to remove saved etag if any, CopyObject calculates a new one.
+	delete(defaultMeta, "etag")
```
Comment (balamurugana) [inline cmd/object-handlers.go:292]: passed `defaultMeta` is a pointer.  Do we want to modify the source?

File: cmd/object-handlers.go [hunk 5]
```diff
@@ -576,7 +573,7 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 			sha256sum = r.Header.Get("X-Amz-Content-Sha256")
 		}
 		// Create object.
-		objInfo, err = objectAPI.PutObject(bucket, object, NewHashReader(r.Body, size, metadata["etag"], sha256sum), metadata)
+		objInfo, err = objectAPI.PutObject(bucket, object, NewHashReader(r.Body, size, md5sum, sha256sum), metadata)
```
Comment (balamurugana) [inline cmd/object-handlers.go]: These changes are just diff increaser.  You can optimize things is different PR IMO

File: cmd/object-api-interface.go [hunk 6]
```diff
@@ -114,29 +130,25 @@ func (r *HashReader) Read(p []byte) (n int, err error) {
...
-func (r *HashReader) MD5() []byte { return r.md5Hash.Sum(nil) }
-
-// Verify verifies if the computed MD5 sum - and SHA256 sum - are
+// verify verifies if the computed MD5 sum - and SHA256 sum - are
 // equal to the ones specified when creating the HashReader.
-func (r *HashReader) Verify() error {
+func (r *HashReader) verify() error {
 	if r.sha256Hash != nil {
-		sha256Sum, err := hex.DecodeString(r.sha256Sum)
+		sha256Sum, err := hex.DecodeString(r.SHA256Sum)
```
Comment (balamurugana) [inline cmd/object-api-interface.go]: This check should be in `NewHashReader()` function.

File: cmd/object-api-interface.go [hunk 7]
```diff
@@ -114,29 +130,25 @@ func (r *HashReader) Read(p []byte) (n int, err error) {
...
-		if !bytes.Equal(sha256Sum, r.sha256Hash.Sum(nil)) {
-			return errContentSHA256Mismatch
+		if sum := r.sha256Hash.Sum(nil); !bytes.Equal(sha256Sum, sum) {
+			return SHA256Mismatch{r.SHA256Sum, hex.EncodeToString(sum)}
 		}
 	}
-	if r.md5Hash != nil && r.md5Sum != "" {
-		md5Sum, err := hex.DecodeString(r.md5Sum)
+	if r.md5Hash != nil {
+		md5Sum, err := hex.DecodeString(r.MD5Sum)
```
Comment (balamurugana) [inline cmd/object-api-interface.go]: same as above

File: cmd/object-api-interface.go [hunk 8]
```diff
@@ -67,30 +67,35 @@ type ObjectLayer interface {
 // HashReader writes what it reads from an io.Reader to an MD5 and SHA256 hash.Hash.
 // HashReader verifies that the content of the io.Reader matches the expected checksums.
 type HashReader struct {
```
Comment (balamurugana) [inline cmd/object-api-interface.go]: Move out `HashReader` from `object-api-interface.go` for better code wake through, may be a package if file list is limited by go

File: cmd/object-api-interface.go [hunk 9]
```diff
@@ -100,12 +105,23 @@ func (r *HashReader) Read(p []byte) (n int, err error) {
 	if err != nil && err != io.EOF {
```
Comment (balamurugana) [inline cmd/object-api-interface.go]: you would need to handle `io.UnexpectedEOF` something like
```go
if err == nil || err == io.UnexpectedEOF {
 // update md5 and sha256 hashers
}

if err == io.EOF || err == io.UnexpectedEOF {
//  call HashReader.Verify()
}

return
```

File: cmd/gateway-s3.go [hunk 10]
```diff
@@ -344,15 +344,14 @@ func (l *s3Objects) GetObjectInfo(bucket string, object string) (objInfo ObjectI
 
 // PutObject creates a new object with the incoming data,
 func (l *s3Objects) PutObject(bucket string, object string, data *HashReader, metadata map[string]string) (objInfo ObjectInfo, err error) {
-	sha256sumBytes, err := hex.DecodeString(data.sha256Sum)
+	sha256sumBytes, err := hex.DecodeString(data.SHA256Sum)
```
Comment (balamurugana) [inline cmd/gateway-s3.go]: I believe SHA256 and MD5 hashes are required to be string with fixed encoding.  You could have respective methods in `HashReader` than public fields.

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: cmd/xl-v1-object.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/xl-v1-object.go]: We can revert this.

File: cmd/xl-v1-object.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/xl-v1-object.go]: Its not needed and incorrect multipart Object copied to new location would validate a wrong ETag and will fail.

File: cmd/test-utils_test.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/test-utils_test.go:1987]: Removing..

File: cmd/object-handlers.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/object-handlers.go]: This is not unrelated @balamurugana - we shouldn't use "etag" in this manner and also this change comes with the change we need to make under HashReader

File: cmd/object-handlers.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/object-handlers.go:292]: Yes we need to remove "etag" we can make it safer.

File: cmd/object-api-interface.go (hunk 9 above)
Comment (harshavardhana) [reply cmd/object-api-interface.go]: Read() never returns io.UnexpectedEOF @balamurugana - UnexpectedEOF is only a higher level construct.

File: cmd/gateway-s3.go (hunk 10 above)
Comment (harshavardhana) [reply cmd/gateway-s3.go]: This change be done..

File: cmd/object-api-interface.go (hunk 8 above)
Comment (harshavardhana) [reply cmd/object-api-interface.go]: This is unrelated to this PR, but i can do that..

File: cmd/object-api-interface.go (hunk 6 above)
Comment (harshavardhana) [reply cmd/object-api-interface.go]: Yeah NewHashReader is not implemented in this manner.. @balamurugana that would require a larger change.

File: cmd/object-api-interface.go (hunk 8 above)
Comment (harshavardhana) [reply cmd/object-api-interface.go]: Also this change would make the PR much bigger.

File: cmd/object-api-interface.go (hunk 6 above)
Comment (balamurugana) [reply cmd/object-api-interface.go]: I guess its related fix only because `HashReader` does verification now which changes the usability.

File: cmd/object-handlers.go (hunk 4 above)
Comment (balamurugana) [reply cmd/object-handlers.go:292]: what do you mean by safer?  what is the problem with passed ETag?

File: cmd/object-handlers.go (hunk 5 above)
Comment (balamurugana) [reply cmd/object-handlers.go]: No.  In the above `md5sum = metadata["etag"]`. you are just optimizing map access.

File: cmd/object-handlers.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/object-handlers.go:292]: you shouldn't pass the "etag" in CopyObject @balamurugana

File: cmd/object-handlers.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/object-handlers.go]: no there is no such thing as "etag" saved in `metadata` it is only saved internally at ObjectLayer. There is no reason to add it prematurely.  You have to see `md5sum` is never set to a map. Also taking "etag" as the md5sum value is incorrect from readability point of view.

File: pkg/hashreader/hashreader.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+	md5sum, sha256sum []byte // Byte values of md5sum, sha256sum of client sent values.
+
+	src                 io.Reader
+	size                int64
+	md5Hash, sha256Hash hash.Hash
+}
+
+// New returns a new HashReader computing the MD5 sum and SHA256 sum
+// (if set) of the provided io.Reader.
+func New(src io.Reader, size int64, md5Hex, sha256Hex string) (*HashReader, error) {
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: I think size can be omitted i.e. consumer needs to restrict by size using `io.LimitReader`

File: pkg/hashreader/hashreader.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
+package hashreader
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: I think this package can be named `hasher` and having a public struct `Reader`

File: pkg/hashreader/hashreader.go [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+	"io"
+)
+
+// HashReader verifies that the content of the io.Reader matches the expected checksums.
+type HashReader struct {
+	md5sum, sha256sum []byte // Byte values of md5sum, sha256sum of client sent values.
+
+	src                 io.Reader
+	size                int64
+	md5Hash, sha256Hash hash.Hash
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: I think checksum and hasher need to be grouped i.e. md5sum and md5hash grouped together.  Also I personally like `md5Hasher` over `md5Hash`

File: pkg/hashreader/hashreader.go [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+	sha256sum, err := hex.DecodeString(sha256Hex)
+	if err != nil {
+		return nil, err
+	}
+	md5sum, err := hex.DecodeString(md5Hex)
+	if err != nil {
+		return nil, err
+	}
+	if hr, ok := src.(*HashReader); ok {
+		return hr, nil
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: As we found given md5/sha256 are valid and if `src` is an `HashReader` I guess we need to error out about nesting is not allowed.  This way we can avoid confusion i.e. user might think newly provided checksums are used in the returned `HashReader`

File: pkg/hashreader/hashreader.go [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+		sha256sum:  sha256sum,
+		src:        io.LimitReader(src, size),
+		size:       size,
+		md5Hash:    md5Hash,
+		sha256Hash: sha256Hash,
+	}, nil
+}
+
+func (r *HashReader) Read(p []byte) (n int, err error) {
+	n, err = r.src.Read(p)
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: `n > 0` may be true in `err != nil`.  I think we need to fix this case as well.

File: pkg/hashreader/hashreader.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+		r.sha256Hash.Write(p[:n])
+	}
+
+	// At io.EOF verify if the checksums are right.
+	if err == io.EOF {
+		cerr := r.Verify()
+		if cerr != nil {
+			return 0, cerr
+		}
+		return n, err
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: this return statement is not required

File: pkg/hashreader/hashreader.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// SHA256Hex returns hex sha256 value.
+func (r *HashReader) SHA256Hex() string {
+	return hex.EncodeToString(r.sha256sum)
+}
+
+// Verify verifies if the computed MD5 sum and SHA256 sum are
+// equal to the ones specified when creating the HashReader.
+func (r *HashReader) Verify() error {
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: Do we need this method as public?  There is no meaning someone calls in the middle

File: pkg/hashreader/hashreader.go [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+	return r.md5sum
+}
+
+// SHA256 - returns byte sha256 value
+func (r *HashReader) SHA256() []byte {
+	return r.sha256sum
+}
+
+// MD5Hex returns hex md5 value.
+func (r *HashReader) MD5Hex() string {
```
Comment (balamurugana) [inline pkg/hashreader/hashreader.go]: I think `MD5HexString()` conveys more meaning

File: pkg/hashreader/hashreader.go (hunk 12 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: Will keep it as hashtee instead hasher means closer to hash.Hash which is not the case here.

File: pkg/hashreader/hashreader.go (hunk 15 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: The case doesn't need to be handled if Read returns any other error other than io.EOF or nil then its a catastrophic.

File: pkg/hashreader/hashreader.go (hunk 17 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: This is still needed there is a bug in erasure code which is not fixed yet.

File: pkg/hashreader/hashreader.go (hunk 13 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: hasher is not go style here we just keep it similar to hash.Hash

File: pkg/hashreader/hashreader.go (hunk 14 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: Yes good idea.

File: pkg/hashreader/hashreader.go (hunk 11 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: Size is needed since we do not pass Size param anymore into our object interface.

File: pkg/hashreader/hashreader.go (hunk 18 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: done

File: pkg/hashreader/hashreader.go (hunk 15 above)
Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: > An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF. 

This line explains this.

Comment (balamurugana) [reply pkg/hashreader/hashreader.go]: Here is the text from https://golang.org/pkg/io/#Reader
```
When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF. 
```

Comment (harshavardhana) [reply pkg/hashreader/hashreader.go]: Correct but we don't need to do anything @balamurugana we already handle this.. it doesn't matter if n > 0 and err is anything other than io.EOF or nil it is a catastrophic error.  We only need to handle things for correct cases which are only two i.e either io.EOF or nil. 

So Reader implementations if n > 0 always `return either err == EOF or err == nil`  any other error is considered catastrophic and should be quickly failed.

**Comment by harshavardhana** [comment]
@balamurugana @donatello please take a look at the updated PR,  package is `pkg/hashtee` with tests. 

@balamurugana i didn't make the change of moving the size param as the caller responsibility of providing the reader, as there were lot of complications in our tests and certain part of the code base. I have left the change for another time perhaps.

**Comment by harshavardhana** [comment]
hashtee is now just `hash` with New renamed to NewReader.  @balamurugana @donatello please take a look.

**Review by donatello — APPROVED** [review]

**Review by balamurugana — APPROVED** [review]
LGTM

**Review by balamurugana — CHANGES_REQUESTED** [review]
CI fails.  Please fix it.

**Review by balamurugana — APPROVED** [review]
LGTM
